File: apps/ios/Sources/OlloCrypto/DeviceRoster.swift
import CryptoKit
import Foundation

/// Hash of the live device roster. Includes device ids so a restored extra
/// device with the same identity keys is still visible. Matches
/// `packages/crypto/src/safety.ts` `deviceRosterHash`.
public enum DeviceRoster {
    public struct Device: Sendable {
        public var deviceId: String
        public var identityX25519: Data

        public init(deviceId: String, identityX25519: Data) {
            self.deviceId = deviceId
            self.identityX25519 = identityX25519
        }
    }

    public enum Decision: String, Sendable {
        case new
        case unchanged
        case changed
    }

    public static func hash(_ devices: [Device]) -> String {
        let sorted = devices.sorted { $0.deviceId < $1.deviceId }
        var out = Data("ollo-roster-v1".utf8)
        for d in sorted {
            out.append(Data(d.deviceId.utf8))
            out.append(0)
            out.append(d.identityX25519)
            out.append(0)
        }
        return SHA256.hash(data: out).map { String(format: "%02x", $0) }.joined()
    }

    public static func note(previous: String?, next: String) -> Decision {
        guard let previous else { return .new }
        return previous == next ? .unchanged : .changed
    }
}

File: apps/android/core/crypto/src/main/java/app/ollo/crypto/DeviceRoster.kt
package app.ollo.crypto

import java.io.ByteArrayOutputStream
import java.security.MessageDigest

/**
 * Hash of the live device roster. Includes device ids so a restored extra
 * device with the same identity keys is still visible. Must match
 * `packages/crypto/src/safety.ts` `deviceRosterHash`.
 */
object DeviceRoster {
    data class Device(val deviceId: String, val identityX25519: ByteArray)

    enum class Decision { New, Unchanged, Changed }

    fun hash(devices: List<Device>): String {
        val sorted = devices.sortedBy { it.deviceId }
        val out = ByteArrayOutputStream()
        out.write("ollo-roster-v1".toByteArray(Charsets.UTF_8))
        for (d in sorted) {
            out.write(d.deviceId.toByteArray(Charsets.UTF_8))
            out.write(0)
            out.write(d.identityX25519)
            out.write(0)
        }
        val digest = MessageDigest.getInstance("SHA-256").digest(out.toByteArray())
        return digest.joinToString("") { b -> "%02x".format(b.toInt() and 0xff) }
    }

    fun note(previous: String?, next: String): Decision {
        if (previous == null) return Decision.New
        return if (previous == next) Decision.Unchanged else Decision.Changed
    }
}
